File: solutions/python/2020/day02.py
from common2020 import str_list, print_ans


class PasswordAndPolicy:
    def __init__(self, min, max, letter, password):
        self.min = min
        self.max = max
        self.letter = letter
        self.password = password


def build_PasswordAndPolicy(password_line) -> PasswordAndPolicy:
    rule, letter, password = password_line.split()

    min, max = rule.split("-")
    letter = letter.strip(":")

    return PasswordAndPolicy(int(min), int(max), letter, password)


def password_is_valid_pt1(obj):
    return obj.min <= obj.password.count(obj.letter) <= obj.max


def password_is_valid_pt2(obj):
    first_char = obj.password[obj.min - 1]
    second_char = obj.password[obj.max - 1]

    return (first_char == obj.letter or second_char == obj.letter) \
        and first_char != second_char


def main():
    passwords = str_list(2)
    pt1 = 0
    pt2 = 0

    for password in passwords:
        password_and_policy_obj = build_PasswordAndPolicy(password)

        if password_is_valid_pt1(password_and_policy_obj): pt1 += 1
        if password_is_valid_pt2(password_and_policy_obj): pt2 += 1
        
    print_ans(pt1, pt2)

main()

File: solutions/2020/day02.ts
import { input, print } from "common";

interface PasswordPolicy {
  min: number;
  max: number;
  char: string;
  password: string;
}

function passwordPolicy(input: string[]): PasswordPolicy {
  return {
    min: parseInt(input[0]),
    max: parseInt(input[1]),
    char: input[2],
    password: input[3],
  };
}

function puzzleInput(): PasswordPolicy[] {
  return input(2020, 2)
    .split("\n")
    .map((raw) =>
      passwordPolicy(
        raw.replace(":", "")
          .replace("-", " ")
          .split(" "),
      )
    );
}

function count(str: string, char: string): number {
  return str.split(char).length - 1;
}

function partOne(): number {
  return puzzleInput().filter((pw) =>
    count(pw.password, pw.char) >= pw.min &&
    count(pw.password, pw.char) <= pw.max
  ).length;
}

function partTwo(): number {
  return puzzleInput().filter((pw) =>
    pw.password.at(pw.min - 1) !=
      pw.password.at(pw.max - 1) &&
    (pw.password.at(pw.min - 1) == pw.char ||
      pw.password.at(pw.max - 1) == pw.char)
  ).length;
}

print(partOne(), partTwo());
